src/codewhisperer/service/keyStrokeHandler.ts

```diff
@@ -215,3 +169,106 @@ export class KeyStrokeHandler {
         }
     }
 }
+
+export abstract class DocumentChangedType {
+    protected contentChanges: ReadonlyArray<vscode.TextDocumentContentChangeEvent>
+
+    constructor(contentChanges: ReadonlyArray<vscode.TextDocumentContentChangeEvent>) {
+        this.contentChanges = contentChanges
+    }
+
+    abstract checkChangeSource(): DocumentChangedSource
+
+    // Enter key should always start with ONE '\n' and potentially following spaces due to IDE reformat
+    static isEnterKey(str: string): boolean {
+        if (str.length === 0) return false
+        return str[0] === '\n' && str.substring(1).trim().length === 0
+    }
+
+    // Tab should consist of space char only ' ' and the length % tabSize should be 0
+    static isTabKey(str: string): boolean {
+        const tabSize = getTabSizeSetting()
+        if (str.length % tabSize === 0 && str.trim() === '') {
+            return true
+        }
+        return false
+    }
+
+    static isUserTypingSpecialChar(str: string): boolean {
+        return ['(', '()', '[', '[]', '{', '{}', ':'].includes(str)
+    }
+
+    static isSingleLine(str: string): boolean {
```

Comment:
Looks like we're calling all these functions with `this.contentChanges[0].text`
Maybe they could be instance methods instead?